src/lib/libdylink.js

```diff
@@ -932,6 +947,59 @@ var LibraryDylink = {
     return dso;
   },
 
+#if FILESYSTEM
+  $findLibraryFS__deps: [
+    '$replaceORIGIN',
+    '_emscripten_find_dylib',
+    '$withStackSave',
+    '$stackAlloc',
+    '$lengthBytesUTF8',
+    '$stringToUTF8OnStack',
+    '$stringToUTF8',
+    '$FS',
+    '$PATH',
+#if WASMFS
+    '_wasmfs_identify',
+    '_wasmfs_read_file',
+#endif
+  ],
+  $findLibraryFS: (libName, rpath) => {
+    // If we're preloading a dynamic library, the runtime is not ready to call
+    // __wasmfs_identify or __emscripten_find_dylib. So just quit out.
+    //
+    // This means that DT_NEEDED for the main module and transitive dependencies
+    // of it won't work with this code path. Similarly, it means that calling
+    // loadDynamicLibrary in a preRun hook can't use this code path.
+    if (!runtimeInitialized) {
+      return undefined;
+    }
+    if (PATH.isAbs(libName)) {
+#if WASMFS
+      var result = withStackSave(() => __wasmfs_identify(stringToUTF8OnStack(libName)));
+      return result === {{{ cDefs.EEXIST }}} ? libName : undefined;
+#else
+      try {
+        FS.lookupPath(libName);
+        return libName;
+      } catch (e) {
+        return undefined;
+      }
+#endif
+    }
+    var rpathResolved = (rpath?.paths || []).map((p) => replaceORIGIN(rpath?.parentLibPath, p));
+    return withStackSave(() => {
+      // In dylink.c we use: `char buf[2*NAME_MAX+2];` and NAME_MAX is 255.
+      // So we use the same size here.
+      var bufSize = 2*255 + 2;
+      var buf = stackAlloc(bufSize);
+      var rpathC = stringToUTF8OnStack(rpathResolved.join(':'));
+      var libNameC = stringToUTF8OnStack(libName);
+      var resLibNameC = __emscripten_find_dylib(buf, rpathC, libNameC, bufSize);
+      return (resLibNameC !== libNameC) ? UTF8ToString(resLibNameC) : undefined;
```

Comment:
Should __emscripten_find_dylib return NULL when it can't find the file?   Then you don't need this pointer comparison?

Reply:
Oh I guess its used internally so maybe we don't want to change the interface.

Reply:
I'll have to check what the other callers of `_emscripten_find_dylib` expect.

Reply:
Yeah I guess I can add a wrapper function for calling from C that keeps the current behavior so that JS can get what it wants and C can get what it wants too.

Reply:
It's a bit annoying that in C, `x || y` isn't the same as `x ? x : y` like it is in JS.